### ISSUE TITLE
Build instructions (coq 8.11.1, VST 2.5, Arch linux)

### ISSUE THREAD

**Issue by mcelrath** [ElementsProject/simplicity#52]
I had quite a bit of trouble building this. I'm none to happy about installing an entirely new package manager with root permissions (nix). I see that v2.5 of VST has been released. v2.4 (nor v2.5) doesn't compile with coq 8.11. I ended up with coq 8.11.1 installed via opam, which compiles their master github branch successfully.

One important contribution below is that VST doesn't compile `sha/functional_prog.vo` by default, so must be done manually.

Anyway, here are some build instructions (Arch Linux):
```
sudo pacman -S opam
opam init
. ~/.bash_profile
opam repo add coq-released https://coq.inria.fr/opam/released
opam install -j`nproc` coq coq-flocq
git clone git@github.com:PrincetonUniversity/VST.git
cd VST
make -j`nproc`
make sha/functional_prog.vo         # This does NOT get built by default!
mkdir -p ${XDG_DATA_HOME:-$HOME/.local/share}/coq/VST
cp -r msl sepcomp veric floyd ${XDG_DATA_HOME:-$HOME/.local/share}/coq/VST
cp -r compcert sha ${XDG_DATA_HOME:-$HOME/.local/share}/coq
cd ../Coq
coq_makefile -f _CoqProject -o Makefile
make
```

And now I don't know what to do with myself. An `examples/` directory would be great! ;-)

**Comment by ProofOfKeags** [comment]
fwiw, there's a workaround for using nix without root. https://github.com/NixOS/nix/issues/585. I haven't personally tried this, maybe you have so if this is off base, let me know. The advantage of nix is that it not only is capable of building the Coq project but the Haskell one as well and I _think_ but haven't verified that it can do the C project too.

**Comment by roconnor-blockstream** [comment]
Thanks for the build instructions Bob!  As you observed, a few weeks ago VST's master branch migrated to CompCert 3.7 and Coq 8.11.  I'm expecting a new release of VST any day now that incorporates these changes, and I plan to update the Nix builds at that point in time.  When that happens, I'll also see about adding your build instructions to the README file.

As for what to do now that you've managed to build the Simplicity Coq project, in principle you can have a look at [Word.v](https://github.com/ElementsProject/simplicity/blob/master/Coq/Simplicity/Word.v), and, in particular, the [Definitions](https://github.com/ElementsProject/simplicity/blob/3eaaac07ac2e5e9d1340fd1f9527da197a5a4669/Coq/Simplicity/Word.v#L266-L472) and [Specifications](https://github.com/ElementsProject/simplicity/blob/3eaaac07ac2e5e9d1340fd1f9527da197a5a4669/Coq/Simplicity/Word.v#L474-L825) sections which define arithmetic and bitwise operations, provide Coq specifications of those functions, and prove the correctness of the Simplicity definitions.  These are best browsed using `coqide` (or any alternative interface such as [Proof General](https://proofgeneral.github.io/)).  Similar work is done in the [SHA256 module](https://github.com/ElementsProject/simplicity/blob/master/Coq/Simplicity/SHA256.v).

The next step for progressing along these lines would be to copy the Simplicity elliptic curve function definitions from [Haskell](https://github.com/ElementsProject/simplicity/blob/3eaaac07ac2e5e9d1340fd1f9527da197a5a4669/Haskell/Core/Simplicity/Programs/LibSecp256k1.hs#L452-L735) and their [Haskell specifications](https://github.com/ElementsProject/simplicity/blob/3eaaac07ac2e5e9d1340fd1f9527da197a5a4669/Haskell/Core/Simplicity/LibSecp256k1/Spec.hs) to Coq. For elliptic curves there are two levels of specifications to be implemented.  The first level is similar to the [Haskell specifications](https://github.com/ElementsProject/simplicity/blob/3eaaac07ac2e5e9d1340fd1f9527da197a5a4669/Haskell/Core/Simplicity/LibSecp256k1/Spec.hs) which specifies precisely the data representations produced by the Simplicity programs.  This level would be used to connect the Simplicity implementations with the jetted C implementations.  A second level of specification would add mathematical abstractions of field values and elliptic curve points.  This is the level where we would prove the elliptic curve group laws, and show that the group has prime order.  (From here we might  reasoning about the Schnorr security reductions to the difficulty of the discrete log problem, or assumptions about pre-image resistance of the hash function, but that is a whole other ball game.)

If someone is interested in pursing that development, I'm happy to open an issue for it.  However it almost certainly requires some significant prior experience working with Coq.

**Comment by roconnor-blockstream** [comment]
VST is now part of Coq's OPAM @ http://coq.io/opam/coq-vst.2.6.html.  I expect this will simplify these build instructions for Arch linux.  I'll try to enlist @apoelstra's help in refining these instruction.  However if anyone else happens to want to give it a try, I would certainly be grateful. ^_^

**Comment by roconnor-blockstream** [comment]
I've updated the non-nix build instruction in PR #63.

**Comment by roconnor-blockstream** [comment]
Fixed in PR #63.
